lib/puppet/transaction/resource_harness.rb

```diff
@@ -120,7 +120,7 @@ def sync_if_needed(param, context)
     do_audit = context.audited_params.include?(param.name)
 
     begin
-      if param.should && !param.safe_insync?(current_value)
+      if !(param.should.nil? || param.safe_insync?(current_value))
```

Comment:
We have to be very careful when changing the resource harness. If I'm understanding correctly, you're differentiating between `should` returning `false` vs `nil`. So if the desired/should value is `false`, then we still want to call `safe_insync?` to see if the current value is also `false`? While that makes sense, I'm not sure that's safe to do. For example in PUP-8442 we tried to change how `false` parameters are handled, but that caused the vcsrepo module to set `force => true` deleting workspaces. See https://tickets.puppetlabs.com/browse/PUP-8442?focusedCommentId=537367&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-537367, and we had to revert that change..

Reply:
Yes, this change will allow a property with `false` value to be passed down for further checks/changes. I don't see how `Puppet::Property::Boolean` will work as a property type otherwise. I checked all `should` methods from puppet code and it should not cause a problem for them.

Maybe allowing `false` only for `Puppet::Property::Boolean` will be safer, even it will not be clean.

Reply:
Historically modules have used different approaches to work around that limitation in puppet. For example the [`atboot` zone property](https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/type/zone.rb#L186-L190), which works since `:false` is truthy. The mount and yumrepo types do something similar, munging/normalizing the value into a symbol like :yes/:no, etc.

Reply:
I changed to using `:true`/`:false` as property values, no change in resource harness is needed.